src/redprl/signature_mini.sml

```diff
@@ -0,0 +1,165 @@
+     sort : sort,
+     spec : jdg option,
+     state : Lcf.jdg Lcf.state}
+
+  type src_catjdg = ast RedPrlCategoricalJudgment.jdg
+  type src_seqhyp = string * src_catjdg
+  type src_sequent = src_seqhyp list * src_catjdg
+  type src_genjdg = ((string * psort) list * (string * sort) list) * src_sequent
+  type src_rulespec = src_genjdg list * src_sequent
+
+  datatype src_decl =
+      DEF of {arguments : string arguments, params : string params, sort : sort, definiens : ast}
+    | THM of {arguments : string arguments, params : string params, goal : src_sequent, script : ast}
+    | RULE of {arguments : string arguments, params : string params, spec : src_rulespec, script : ast}
+    | TAC of {arguments : string arguments, params : string params, script : ast}
+
+  datatype 'opid cmd =
+      PRINT of 'opid
+    | EXTRACT of 'opid
+
+  type src_cmd = src_opid cmd
+
+  datatype src_elt =
+      DECL of string * src_decl * Pos.t
+    | CMD of src_cmd * Pos.t
+
+  (* elaborated declarations *)
+  datatype elab_decl =
+      EDEF of entry
+    | ECMD of opid cmd
+
+  structure Telescope = Telescope (StringAbtSymbol)
+  structure ETelescope = Telescope (Tm.Sym)
+  structure NameEnv = AstToAbt.NameEnv
+
+  (* A signature / [sign] is a telescope of declarations. *)
+  type src_sign = (src_decl * Pos.t option) Telescope.telescope
+
+  (* An elaborated signature is a telescope of definitions. *)
+  type elab_sign = elab_decl ElabMonad.t ETelescope.telescope
+
+  type sign =
+    {sourceSign : src_sign,
+     elabSign : elab_sign,
+     nameEnv : Tm.symbol NameEnv.dict}
+
+  structure E = ElabMonadUtil (ElabMonad)
+  fun lookup ({elabSign, ...} : sign) opid =
+    case E.run (ETelescope.lookup elabSign opid) of
+        SOME (EDEF defn) => defn
+      | _ => raise Fail "Elaboration failed"
+
+  fun unifyCustomOperator (entry : entry) (ps : Tm.param list) (es : abt Tm.bview list) : Tm.metaenv * Tm.symenv =
+    let
+      val {params, arguments, ...} = entry
+      val srho = ListPair.foldl (fn ((u, _), p, ctx) => Sym.Ctx.insert ctx u p) Sym.Ctx.empty (params, ps)
+      val mrho = ListPair.foldl (fn ((x, vl), e, ctx) => Metavar.Ctx.insert ctx x (Tm.checkb (e, vl))) Metavar.Ctx.empty (arguments, es)
+    in
+      (mrho, srho)
+    end
+
+
+  local 
+    open RedPrlOpData Tm RedPrlSequent
+    infix $ \ >>
+
+    (* Observe that hypotheses have a dual nature: they are both symbols and variables. When reviving a proof state,
+      we have to rename hypotheses in *both* their moments. This routine constructs the appropriate substitution
+      for hypotheses qua variables. *)
+    fun hypothesisRenaming (entry : entry) (ps : Tm.param list) : Tm.varenv = 
+      let
+        fun handleHyp ((u, psort), ptm, ctx) = 
+          case psort of 
+            RedPrlParamData.HYP =>
+              let
+                val RedPrlParameterTerm.VAR v = ptm
+              in
+                Var.Ctx.insert ctx u (Tm.check (Tm.`v, RedPrlOpData.EXP))
```

Comment:
This is not quite right; we need to somehow find out what judgment the hypothesis is for, and use its synthesis instead of `EXP`.